### PR TITLE
feat(teams): Add team slugs

### PR DESCRIPTION
**Pull request by cjdenio**
This PR adds a unique `slug` field to every team.

Things to note:

- `name` is now optional (the UI should treat `slug` as the team's name if `name` is not present)
- `slug` can only contain lowercase letters and hyphens for now
- Auto-created personal teams have a slug based on their username (e.g. `caleb-denio-3977`)
- wahoo

### CODE REVIEW

File: src/api/auth.rs [hunk 1]
```diff
@@ -88,7 +89,8 @@ pub async fn code(conn: DbConn, code: &str, cookies: &CookieJar<'_>) -> Result<R
                     // Create the user's personal team
                     let team = diesel::insert_into(teams)
                         .values(&NewTeam {
-                            name: format!("{}'s team", info.name),
+                            name: Some(format!("{}'s team", info.name)),
+                            slug: Generator::default().next().unwrap(),
```
Comment (anirudhb) [inline src/api/auth.rs]: Can you make the slug the username instead as it makes it easier later on (as only to the user we have the global namespace, not in the DB internally)

**Review by anirudhb — COMMENTED** [review]
Nit, lgtm

File: src/api/auth.rs [hunk 2]
```diff
@@ -88,7 +88,8 @@ pub async fn code(conn: DbConn, code: &str, cookies: &CookieJar<'_>) -> Result<R
                     // Create the user's personal team
                     let team = diesel::insert_into(teams)
                         .values(&NewTeam {
-                            name: format!("{}'s team", info.name),
+                            name: Some(format!("{}'s team", info.name)),
+                            slug: into_slug(&info.name, true),
```
Comment (anirudhb) [inline src/api/auth.rs:92]: Should be based off username, not "real name" field. This should also fix the issue with spaces and such, but we should substitute `_` for `-` in usernames and remove all other characters.

File: src/models/team.rs [hunk 3]
```diff
@@ -1,22 +1,54 @@
...
+
+/// Converts any string to a team-compatible slug
+pub fn into_slug(text: &str, randomize: bool) -> String {
+    lazy_static! {
+        static ref INVALID_REGEX: Regex = Regex::new("[^a-z0-9 ]").unwrap();
+    }
+
+    let slug = INVALID_REGEX
+        .replace_all(&text.trim().to_lowercase(), "")
+        .replace(" ", "-");
```
Comment (anirudhb) [inline src/models/team.rs]: It's probably better to restrict the invalid regex and replace the space before deleting invalid characters.

File: src/api/auth.rs (hunk 2 above)
Comment (cjdenio) [reply src/api/auth.rs:92]: Slack only provides a display name and a "real" name; unique usernames (like `cjdenio`) were deprecated before and don't exist in the new OpenID Connect API

File: src/models/team.rs [hunk 4]
```diff
@@ -1,22 +1,54 @@
 use crate::schema::teams;
 use chrono::NaiveDateTime;
+use regex::Regex;
 use serde::{Deserialize, Serialize};
 
+use rand::prelude::*;
+
+lazy_static! {
+    static ref SLUG_REGEX: Regex = Regex::new("^[a-z0-9\\-]+$").unwrap();
```
Comment (cfanoulis) [inline src/models/team.rs:9]: regex should only be a-z0-9_-
```suggestion
    static ref SLUG_REGEX: Regex = Regex::new("^[a-z0-9-_]+$").unwrap();
```

File: src/models/team.rs [hunk 5]
```diff
@@ -1,22 +1,54 @@
...
+}
+
+pub fn validate_slug(slug: &str) -> bool {
+    SLUG_REGEX.is_match(slug)
+}
+
+/// Converts any string to a team-compatible slug
+pub fn into_slug(text: &str, randomize: bool) -> String {
+    lazy_static! {
+        static ref INVALID_REGEX: Regex = Regex::new("[^a-z0-9\\-]").unwrap();
```
Comment (cfanoulis) [inline src/models/team.rs:40]: regex should only be a-z0-9_-
```suggestion
        static ref SLUG_REGEX: Regex = Regex::new("^[a-z0-9-_]+$").unwrap();
```

File: src/models/token.rs [hunk 6]
```diff
@@ -4,7 +4,6 @@ use chrono::NaiveDateTime;
 
 #[derive(Debug, Queryable, Identifiable, Associations)]
 #[primary_key(token)]
-#[table_name = "tokens"]
```
Comment (cfanoulis) [inline src/models/token.rs]: seperate PR please

File: src/api/teams.rs [hunk 7]
```diff
@@ -12,6 +12,10 @@ use crate::{
 
 #[post("/teams", data = "<team>")]
 pub async fn create(user: User, team: Json<NewTeam>, conn: DbConn) -> Result<Json<Team>, Status> {
+    if !validate_slug(&team.slug) {
+        return Err(Status::UnprocessableEntity);
```
Comment (cfanoulis) [inline src/api/teams.rs:16]: this should convert the slug to an acceptable version

**Review by cfanoulis — CHANGES_REQUESTED** [review]
Regex stuff

File: src/api/teams.rs (hunk 7 above)
Comment (cfanoulis) [reply src/api/teams.rs:16]: or rather, add an option?

File: src/models/team.rs (hunk 4 above)
Comment (cjdenio) [reply src/models/team.rs:9]: Why? The current regex matches lowercase characters + numbers + hyphens

File: src/api/teams.rs (hunk 7 above)
Comment (cjdenio) [reply src/api/teams.rs:16]: that creates an unexpected side effect with API usage; if your parameters are invalid, the server should tell you so

File: src/models/team.rs (hunk 4 above)
Comment (cfanoulis) [reply src/models/team.rs:9]: it actually matches `\` as well which may break it

Comment (cjdenio) [reply src/models/team.rs:9]: It actually doesn't, the `\\` is just an escape character for the `-` (two are needed so Rust interprets it as a literal `\`)

**Review by anirudhb — APPROVED** [review]
wahoo, hopefully nothing goes wrong

**Pull request merged**